packages/styled-components/src/utils/flatten.js

```diff
@@ -9,40 +9,62 @@ import Keyframes from '../models/Keyframes';
 import hyphenate from './hyphenateStyleName';
 import addUnitIfNeeded from './addUnitIfNeeded';
 
+type ObjectToCSSArrayProps = {
+  chunk: Object,
+  executionContext?: Object,
+  prevKey?: string,
+  styleSheet?: Object,
+};
+
 /**
  * It's falsish not falsy because 0 is allowed.
  */
-const isFalsish = chunk => chunk === undefined || chunk === null || chunk === false || chunk === '';
+function isFalsish(chunk: any): boolean {
+  return chunk === undefined || chunk === null || chunk === false || chunk === '';
+}
 
+function objToCssArray(args: ObjectToCSSArrayProps): Array<string | Function> {
+  const rules = [];
+  const keys = Object.keys(args.chunk);
 
-const objToCssArray = (obj: Object, prevKey?: string): Array<string | Function> => {
-  const rules = []
-  const keys = Object.keys(obj)
+  keys.forEach(key => {
+    const value = args.chunk[key];
 
-  keys
-    .forEach(key => {
-      if (!isFalsish(obj[key])) {
-        if (isPlainObject(obj[key])) {
-          rules.push(...objToCssArray(obj[key], key))
+    if (!isFalsish(value)) {
+      if (Array.isArray(value) /* likely usage of `css` tagged template literal */) {
+        /* eslint-disable no-use-before-define */
+        return rules.push(
+          `${hyphenate(key)}: ${flatten(value, args.executionContext, args.styleSheet).join('')};`
```

Comment:
not sure about `.join` part actually.
After this change, this should be acceptable and if it's, then i think this will cause a bug.

```js
styled({
  h1: [{
        color: $(props) => props.color
    }]
})
```

we call `flattern` and will end up in `objToCssArray` with an object `{ color: $(props) => props.color }` which will return

```js
['color:', func, ...]
```
and with `.join` will cause the previous bug #2489

Reply:
actually tested it against your branch 
```diff
+   "h1: color:function color(props) {
+           return props.color;
+         };;",
```

also, one more thing to consider, we still need `objectToCssArray` to returns an array for `isStatic` flag.

my suggesting is to keep `objectToCssArray` context-free and split `flatten` handling of the array which is going to be context-free as well, and reuse it in both.